userspace/falco/app_cmdline_options.cpp

```diff
@@ -0,0 +1,209 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "app_cmdline_options.h"
+#include "configuration.h"
+#include "config_falco.h"
+
+#include <fstream>
+
+namespace falco {
+namespace app {
+
+// Most bool member variables do not need to be set explicitly, as
+// they are bound to command line options that have default
+// values. However, a few options can be ifdef'd out so explicitly
+// initialize their linked variables.
+cmdline_options::cmdline_options()
+	: event_buffer_format(sinsp_evt::PF_NORMAL),
+	  list_plugins(false),
+	  m_cmdline_opts("falco", "Falco - Cloud Native Runtime Security")
+{
+	define();
+}
+
+cmdline_options::~cmdline_options()
+{
+}
+
+bool cmdline_options::parse(int argc, char **argv, std::string &errstr)
+{
+	try {
+		m_cmdline_parsed = m_cmdline_opts.parse(argc, argv);
+	}
+	catch (std::exception &e)
+	{
+		errstr = e.what();
+		return false;
+	}
+
+	// Some options require additional processing/validation
+	std::ifstream conf_stream;
+	if (!conf_filename.empty())
+	{
+		conf_stream.open(conf_filename);
+		if (!conf_stream.is_open())
+		{
+			errstr = std::string("Could not find configuration file at ") + conf_filename;
+			return false;
+		}
+	}
+	else
+	{
+#ifndef BUILD_TYPE_RELEASE
+		conf_stream.open(FALCO_SOURCE_CONF_FILE);
+		if (conf_stream.is_open())
+		{
+			conf_filename = FALCO_SOURCE_CONF_FILE;
+		}
+		else
+#endif
+		{
+			conf_stream.open(FALCO_INSTALL_CONF_FILE);
+			if (conf_stream.is_open())
+			{
+				conf_filename = FALCO_INSTALL_CONF_FILE;
+			}
+			else
+			{
+				// Note we do not return false here. Although there is
+				// no valid config file, some ways of running falco
+				// (e.g. --help, --list) do not need a config file.
+				//
+				// Later, when it comes time to read a config file, if
+				// the filename is empty we exit with an error.
+				conf_filename = "";
+			}
+		}
+	}
+
+	if(m_cmdline_parsed.count("b") > 0)
+	{
+		event_buffer_format = sinsp_evt::PF_BASE64;
+	}
+
+	// Expand any paths provided via -r and fill in rules_filenames
+	if(m_cmdline_parsed.count("r") > 0)
+	{
+		for(auto &path : m_cmdline_parsed["r"].as<std::vector<std::string>>())
+		{
+			falco_configuration::read_rules_file_directory(path, rules_filenames);
+		}
+	}
+
+	// Convert the vectors of enabled/disabled tags into sets to match falco engine API
+	if(m_cmdline_parsed.count("T") > 0)
+	{
+		for(auto &tag : m_cmdline_parsed["T"].as<std::vector<std::string>>())
+		{
+			disabled_rule_tags.insert(tag);
+		}
+	}
+
+	if(m_cmdline_parsed.count("t") > 0)
+	{
+		for(auto &tag : m_cmdline_parsed["t"].as<std::vector<std::string>>())
+		{
+			enabled_rule_tags.insert(tag);
+		}
+	}
+
+	// Some combinations of arguments are not allowed.
+
+	// You can't both disable and enable rules
+	if((disabled_rule_substrings.size() + disabled_rule_tags.size() > 0) &&
```

Comment:
Maybe I'm missing something, but you seem to initialize `disable_rule_substrings` differently. The first is initialized as a single value with `cxxopts::value(disabled_rule_substrings)` even though it's a vector, whereas the second is initialized manually value-by-value [in this method at line 108](https://github.com/falcosecurity/falco/pull/1886/files#diff-e8ce29b2b289869d198bfb7de910426be6c5c226e4ee26cdc9f42eed9c68d787R108). Is there a reason for this difference, or is `disabled_rule_substrings` supposed to be inizialized manually too?

Reply:
cxxopts allows you to bind an option to a vector (https://github.com/jarro2783/cxxopts#stdvectort-values). In this case, specifying the option multiple times appends to the vector. That's the approach taken for disabled_rule_substrings.

disabled_rule_tags was handled differently mostly to leave the falco engine interface unchanged. `falco_engine::enable_rule_by_tag(const std::set<std::string> &tags, ...)` already takes a std::set as its first argument. However, cxxopts only allows binding to vectors. So the instance variable is a set, but after parsing args the values in the vector for the option `T/t` are added to the set.

We can change the falco interface to use a vector for the method `falco_engine::enable_rule_by_tag()`, in which case we could treat disabled_rule_substrings and disabled_rule_tags the same.

Reply:
There's no need for now I think, my only concern was that everything worked as expected. I personally like having tags as a set, since that's kind of a good data structure for them.

Reply:
I agree we can improve it later. Not a blocker for this PR, IMO.